src/Microsoft.Extensions.DependencyInjection.Abstractions/IRequiredServiceAware.cs

```diff
@@ -0,0 +1,24 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+
+namespace Microsoft.Extensions.DependencyInjection
+{
+    /// <summary>
+    /// Implementation which gets invoked on the <see cref="IServiceProvider"/> when
+    /// <see cref="ServiceProviderExtensions.GetRequiredService{T}(IServiceProvider)"/> or
+    /// <see cref="ServiceProviderExtensions.GetRequiredService(IServiceProvider, Type)"/> is called.
+    /// </summary>
+    public interface IRequiredServiceAware
```

Comment:
One difference here is this is a capability, so it's better to look at other optional interfaces to compare apples to apples. This is more like our feature interfaces than it is like the other ones you mention. `IServiceProvider` and `IServiceScopeFactory` are services in the container itself, I'm not sure this should follow that pattern. This is more of an opportunistic downcast to enable `IServiceProvider` implementors to also provide a GetRequiredService method